Support translation file assets

msteiger opened the pull request:
~~For now, this is mainly a request for comments - it's not complete yet.~~ **This PR is now complete.**

It adds functionality to use ID strings in UI widgets that refer to translation files. This IDs are resolved to translated text strings at runtime. The target language, the ID and the translation string can be edited live (see the video https://youtu.be/Ju_-Fg4XZPQ).

A reference to an i18n ID looks like this:

    ${engine:menu#start-singleplayer-game}

The part inside the curly brackets is used in the language files. The first part identifies the project, the part after the hash is the actual ID.
A new asset type is introduced: `Translation`.
The actual translation information is provided by `TranslationSystem` and its projects.

See http://translate.terasology.org/projects/terasology for the web access.

Every project has a language base file (without locale extension). It describes (in English) what is *meant*. For example, it's hardly possible to translate "get" or "take" without semantic context.

**EDIT**: updated after some refactoring

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/229/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/230/
Hooray Jenkins reported success with all tests good!

immortius commented:
http://forum.terasology.org/threads/internationalisation.1028/#post-11201 for background/previous discussion.

The logic in this PR seems reversed. Rather than having elements referencing internationalised string assets (such as ```engine:menu#singleplayer``` from menu_xx.lang), it has internationalised strings being injected into other things. Pros/cons? 

I would suggest at least the the second ```#``` be replaced with a ```.``` in the Ids.

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/231/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/232/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/233/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/234/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/235/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/236/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/239/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/240/
Hooray Jenkins reported success with all tests good!

msteiger commented:
Okay - all done. See the video for a demonstration: https://youtu.be/Ju_-Fg4XZPQ

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/242/
Hooray Jenkins reported success with all tests good!

msteiger commented:
Hmm ... UIElement assets (sometimes!) seem to reload infinitely when the environment is switched to Preview.

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/243/
Hooray Jenkins reported success with all tests good!

immortius commented on engine/src/main/java/org/terasology/engine/TerasologyEngine.java:
```diff
@@ -187,11 +187,15 @@ private void initialize() {
             environmentSwitcher.handleSwitchToGameEnvironment(rootContext);
 
+            // TODO: find the right place to put this
+            rootContext.put(TranslationSystem.class, new TranslationSystemImpl(rootContext));
```
Please put this in a subsystem.

immortius commented on engine/src/main/java/org/terasology/i18n/StandardTranslationProject.java:
```diff
@@ -0,0 +1,83 @@
+        return result;
+    }
+
+    private String translateExact(String key, Locale locale) {
```
Suggest this should return ```Optional<String>```

immortius commented on engine/src/main/java/org/terasology/i18n/TranslationProject.java:
```diff
@@ -0,0 +1,54 @@
+ * Describes a translation project. It performs the actual mapping of ID strings to human-readable text
+ * in a language that is specified through {@link Locale}.
+ */
+public interface TranslationProject {
```
Not sure ```Project``` describes what this class does?

immortius commented on engine/src/main/java/org/terasology/i18n/StandardTranslationProject.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Override
+    public String translate(String key, Locale locale) {
```
Suggest this should return Optional<String>, or return the key if translation fails.

immortius commented on engine/src/main/java/org/terasology/i18n/TranslationSystemImpl.java:
```diff
@@ -0,0 +1,125 @@
+        config = context.get(Config.class).getSystem();
+        AssetManager assetManager = context.get(AssetManager.class);
+
+        Set<ResourceUrn> urns = assetManager.getAvailableAssets(Translation.class);
```
I wonder if this needs to be done upfront, or could be done on demand. All the information needed to discover the relevant asset is available from a translation uri + locale, so rather than loading them all and indexing, could just search for them as needed?

msteiger replied:
We need to know about all locales for a given base URI, because they could be useful for fallback lookups. I prefer to do that once (and update if needed), in particular if translation lookup shappen often.

immortius commented on engine/src/main/java/org/terasology/i18n/TranslationSystemImpl.java:
```diff
@@ -0,0 +1,125 @@
+    public String translate(String id, Locale otherLocale) {
+        int splitPoint = id.indexOf('#');
+        if (splitPoint > 0) {
+            String projName = id.substring(0, splitPoint);
```
You could also load it into an AssetUri, which naturally splits out the different parts?

immortius commented on engine/src/main/java/org/terasology/i18n/assets/Translation.java:
```diff
@@ -0,0 +1,116 @@
+    @Override
+    protected void doReload(TranslationData newData) {
+        Preconditions.checkArgument(newData != null);
+        this.data = newData;
```
Suggest copy of data, or not retaining the data and inlining the details in the asset. Leaves you at risk of the data being changed outside of the proper reload pattern.

immortius commented on engine/src/main/java/org/terasology/rendering/nui/internal/NUIManagerInternal.java:
```diff
@@ -65,7 +70,7 @@
 public class NUIManagerInternal extends BaseComponentSystem implements NUIManager {
     private Logger logger = LoggerFactory.getLogger(NUIManagerInternal.class);
-    private Deque<UIScreenLayer> screens = Queues.newArrayDeque();
+    private LinkedList<UIScreenLayer> screens = new LinkedList<>();
```
Why?

msteiger replied:
The list interface is needed to [reload individual screens](https://github.com/MovingBlocks/Terasology/pull/1896/files#diff-c6ee4291cb4c4614c5d299b8fe8656e7R567).

immortius replied:
Um, is that link wrong? I'm not seeing anything there that requires a linked list.

immortius commented on engine/src/main/java/org/terasology/rendering/nui/internal/NUIManagerInternal.java:
```diff
@@ -551,9 +564,48 @@ public void bindEvent(BindButtonEvent event, EntityRef entity) {
+
     private void prepare(ControlWidget screen) {
         InjectionHelper.inject(screen);
+        translator.updateWidget(screen);
```
Suggest retaining untranslated strings and translating during rendering.

immortius replied:
You really have to for strings obtained through data binding.

immortius commented on engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/videoSettings/LocaleRenderer.java:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * @param context
+     */
+    public LocaleRenderer(Context context) {
```
Change this to take TranslationSystem, not Context.

immortius commented on engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/videoSettings/VideoSettingsScreen.java:
```diff
@@ -56,6 +65,9 @@
     private LocalPlayer localPlayer;
 
+    @In
+    private Context context;
```
Change this to TranslationSystem, not Context.

Cervator commented:
Very cool video. I'm glad to see more work on this topic, but I've probably said that too much already :-)

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/244/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/245/
Hooray Jenkins reported success with all tests good!

msteiger commented:
I've addressed all comments now. Live reloading of UIElements is disabled for now due to issues with the preview screen (see d8d88d0).

Also, only simple UI widgets (labels, buttons, etc) are translated at the moment.

immortius commented on engine/src/main/java/org/terasology/engine/TerasologyEngine.java:
```diff
@@ -315,6 +318,8 @@ private void initAssets() {
                 (AssetFactory<BehaviorTree, BehaviorTreeData>) BehaviorTree::new, false, "behaviors");
         assetTypeManager.registerCoreAssetType(UIElement.class,
                 (AssetFactory<UIElement, UIData>) UIElement::new, "ui");
+        assetTypeManager.registerCoreAssetType(Translation.class,
```
I'ld suggest this should be moved into the I18nSubsystem too (I moved the reset of the asset type initialisations into subsystems in my engine lifecycle rework branch).

immortius commented on engine/src/main/java/org/terasology/i18n/TranslationUri.java:
```diff
@@ -0,0 +1,133 @@
+ * These URIs are always in the form: {@literal <module-name>:<project-name>#<id>}.
+ */
+@API
+public class TranslationUri implements Uri {
```
Do we really need another Uri class? I've been trying to get rid of them. Won't AssetUri do?

immortius replied:
ResourceUrn rather.

immortius commented on engine/src/main/java/org/terasology/rendering/nui/TextLineBuilder.java:
```diff
@@ -54,6 +54,10 @@ public TextLineBuilder(Font font, int maxWidth, int startOffset) {
     }
 
     public void addText(String text) {
+        if (text == null) {
```
Suggest replace with precondition. Null should not be added - it is a bug that should be handled outside of this.

immortius commented on engine/src/main/java/org/terasology/rendering/nui/asset/UIFormat.java:
```diff
@@ -188,6 +192,11 @@ public UIWidget deserialize(JsonElement json, Type typeOfT, JsonDeserializationC
                                 }
                                 field.setValue(element, result);
                             }
+                        } else if (field.getType().equals(String.class)) {
```
Have you considered registering a type adapter for String into gson?

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/247/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/248/
Hooray Jenkins reported success with all tests good!

immortius commented:
Congrats! You have slain the ravenous beast that is my desire for excellence. 

All looks good now, thanks. :)

Cervator commented:
Huzzah! I'd say "Sold!" and merge, but during testing found a readily reproducible crash bug: changing language in-game goes boom:

```
00:36:52.522 [main] WARN  o.t.engine.internal.TimeBase - Delta too great (1349), capping to 1000
00:36:58.169 [main] INFO  o.t.r.n.l.m.v.VideoSettingsScreen - Video Settings: {"pixelFormat":24,"windowPosX":481,"windowPosY":225,"windowWidth":1152,"windowHeight":720,"fullscreen":false,"viewDistance":"moderate","flickeringLight":true,"animateGrass":true,"animateWater":false,"fieldOfView":90.0,"cameraBobbing":true,"renderPlacingBox":true,"blurIntensity":2,"reflectiveWater":false,"vignette":true,"motionBlur":true,"ssao":false,"filmGrain":true,"outline":true,"lightShafts":true,"eyeAdaptation":true,"bloom":true,"dynamicShadows":true,"oculusVrSupport":false,"maxTextureAtlasResolution":4096,"maxChunksUsedForShadowMapping":1024,"shadowMapResolution":1024,"normalMapping":false,"parallaxMapping":false,"dynamicShadowsPcfFiltering":false,"cloudShadows":true,"renderNearest":true,"particleEffectLimit":10,"frameLimit":60,"meshLimit":400,"inscattering":true,"localReflections":false,"vSync":false,"clampLighting":false,"fboScale":100,"dumpShaders":false,"screenshotSize":"normal size","screenshotFormat":"jpg","cameraSettings":{"cameraSetting":"normal"},"debug":{"enabled":false,"stage":"opaque_color","firstPersonElementsHidden":false,"hudHidden":false,"wireframe":false,"renderChunkBoundingBoxes":false,"renderSkeletons":false}}
00:36:58.250 [main] ERROR o.terasology.engine.TerasologyEngine - Uncaught exception, attempting clean game shutdown
java.lang.NullPointerException: null
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:210) ~[guava-18.0.jar:na]
	at org.terasology.assets.AssetType.getAsset(AssetType.java:228) ~[gestalt-asset-core-4.1.2.jar:4.1.2]
	at org.terasology.assets.management.AssetManager.getAsset(AssetManager.java:206) ~[gestalt-asset-core-4.1.2.jar:4.1.2]
	at org.terasology.asset.Assets.get(Assets.java:75) ~[classes/:na]
	at org.terasology.rendering.nui.internal.NUIManagerInternal.pushScreen(NUIManagerInternal.java:225) ~[classes/:na]
	at org.terasology.rendering.nui.internal.NUIManagerInternal.invalidate(NUIManagerInternal.java:581) ~[classes/:na]
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.LocaleBinding.set(LocaleBinding.java:45) ~[classes/:na]
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.LocaleBinding.set(LocaleBinding.java:27) ~[classes/:na]
	at org.terasology.rendering.nui.widgets.UIDropdown.setSelection(UIDropdown.java:156) ~[classes/:na]
	at org.terasology.rendering.nui.widgets.UIDropdown$ItemListener.onMouseClick(UIDropdown.java:172) ~[classes/:na]
	at org.terasology.rendering.nui.internal.CanvasImpl.processMouseClick(CanvasImpl.java:228) ~[classes/:na]
	at org.terasology.rendering.nui.internal.NUIManagerInternal.mouseButtonEvent(NUIManagerInternal.java:479) ~[classes/:na]
	at org.terasology.rendering.nui.internal.NUIManagerInternalMethodAccess.invoke(Unknown Source) ~[reflectasm-1.11.0.jar:na]
	at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:506) ~[classes/:na]
	at org.terasology.entitySystem.event.internal.EventSystemImpl.sendConsumableEvent(EventSystemImpl.java:280) ~[classes/:na]
	at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:259) ~[classes/:na]
	at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:144) ~[classes/:na]
	at org.terasology.input.InputSystem.sendMouseEvent(InputSystem.java:408) ~[classes/:na]
	at org.terasology.input.InputSystem.processMouseInput(InputSystem.java:263) ~[classes/:na]
	at org.terasology.input.InputSystem.update(InputSystem.java:211) ~[classes/:na]
	at org.terasology.engine.modes.StateIngame.handleInput(StateIngame.java:197) ~[classes/:na]
	at org.terasology.engine.subsystem.lwjgl.LwjglInput.postUpdate(LwjglInput.java:55) ~[classes/:na]
	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:421) ~[classes/:na]
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:367) ~[classes/:na]
	at org.terasology.engine.Terasology.main(Terasology.java:153) [classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_45]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_45]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_45]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_45]
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140) [idea_rt.jar:na]
00:36:58.250 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
00:36:58.379 [Chunk-Updater-4] ERROR o.t.u.concurrency.TaskProcessor - Error in thread Chunk-Updater-4
java.lang.NullPointerException: null
	at org.terasology.world.chunks.internal.ChunkImpl.getSunlight(ChunkImpl.java:222) ~[classes/:na]
	at org.terasology.world.chunks.internal.ChunkImpl.getSunlight(ChunkImpl.java:217) ~[classes/:na]
	at org.terasology.world.internal.ChunkViewCoreImpl.getSunlight(ChunkViewCoreImpl.java:140) ~[classes/:na]
	at org.terasology.world.internal.ChunkViewCoreImpl.getSunlight(ChunkViewCoreImpl.java:115) ~[classes/:na]
	at org.terasology.rendering.primitives.ChunkTessellator.calcLightingValuesForVertexPos(ChunkTessellator.java:179) ~[classes/:na]
	at org.terasology.rendering.primitives.ChunkTessellator.generateOptimizedBuffers(ChunkTessellator.java:116) ~[classes/:na]
	at org.terasology.rendering.primitives.ChunkTessellator.generateMesh(ChunkTessellator.java:69) ~[classes/:na]
	at org.terasology.rendering.world.ChunkMeshUpdateManager$ChunkUpdateTask.run(ChunkMeshUpdateManager.java:172) ~[classes/:na]
	at org.terasology.utilities.concurrency.TaskProcessor.run(TaskProcessor.java:50) ~[classes/:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_45]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_45]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
00:36:58.396 [Chunk-Updater-6] ERROR o.t.u.concurrency.TaskProcessor - Error in thread Chunk-Updater-6
java.lang.NullPointerException: null
	at org.terasology.world.chunks.internal.ChunkImpl.getSunlight(ChunkImpl.java:222) ~[classes/:na]
	at org.terasology.world.chunks.internal.ChunkImpl.getSunlight(ChunkImpl.java:217) ~[classes/:na]
	at org.terasology.world.internal.ChunkViewCoreImpl.getSunlight(ChunkViewCoreImpl.java:140) ~[classes/:na]
	at org.terasology.world.internal.ChunkViewCoreImpl.getSunlight(ChunkViewCoreImpl.java:115) ~[classes/:na]
	at org.terasology.rendering.primitives.ChunkTessellator.calcLightingValuesForVertexPos(ChunkTessellator.java:179) ~[classes/:na]
	at org.terasology.rendering.primitives.ChunkTessellator.generateOptimizedBuffers(ChunkTessellator.java:116) ~[classes/:na]
	at org.terasology.rendering.primitives.ChunkTessellator.generateMesh(ChunkTessellator.java:69) ~[classes/:na]
	at org.terasology.rendering.world.ChunkMeshUpdateManager$ChunkUpdateTask.run(ChunkMeshUpdateManager.java:172) ~[classes/:na]
	at org.terasology.utilities.concurrency.TaskProcessor.run(TaskProcessor.java:50) ~[classes/:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_45]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_45]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
00:36:58.463 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
00:36:58.484 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Snapshot created: Writing phase starts
00:36:58.488 [main] INFO  o.t.n.internal.NetworkSystemImpl - Client disconnected: Cervator
00:36:58.506 [main] INFO  o.t.logic.console.ConsoleImpl - [NOTIFICATION] Player "Cervator" has left the game
00:36:58.509 [main] INFO  o.t.n.internal.NetworkSystemImpl - Network shutdown
00:36:59.461 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
00:36:59.462 [main] WARN  org.reflections.Reflections - given scan urls are empty. set urls in the configuration
java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:210)
	at org.terasology.assets.AssetType.getAsset(AssetType.java:228)
	at org.terasology.assets.management.AssetManager.getAsset(AssetManager.java:206)
	at org.terasology.asset.Assets.get(Assets.java:75)
	at org.terasology.rendering.nui.internal.NUIManagerInternal.pushScreen(NUIManagerInternal.java:225)
	at org.terasology.rendering.nui.internal.NUIManagerInternal.invalidate(NUIManagerInternal.java:581)
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.LocaleBinding.set(LocaleBinding.java:45)
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.LocaleBinding.set(LocaleBinding.java:27)
	at org.terasology.rendering.nui.widgets.UIDropdown.setSelection(UIDropdown.java:156)
	at org.terasology.rendering.nui.widgets.UIDropdown$ItemListener.onMouseClick(UIDropdown.java:172)
	at org.terasology.rendering.nui.internal.CanvasImpl.processMouseClick(CanvasImpl.java:228)
	at org.terasology.rendering.nui.internal.NUIManagerInternal.mouseButtonEvent(NUIManagerInternal.java:479)
	at org.terasology.rendering.nui.internal.NUIManagerInternalMethodAccess.invoke(Unknown Source)
	at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:506)
	at org.terasology.entitySystem.event.internal.EventSystemImpl.sendConsumableEvent(EventSystemImpl.java:280)
	at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:259)
	at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:144)
	at org.terasology.input.InputSystem.sendMouseEvent(InputSystem.java:408)
	at org.terasology.input.InputSystem.processMouseInput(InputSystem.java:263)
	at org.terasology.input.InputSystem.update(InputSystem.java:211)
	at org.terasology.engine.modes.StateIngame.handleInput(StateIngame.java:197)
	at org.terasology.engine.subsystem.lwjgl.LwjglInput.postUpdate(LwjglInput.java:55)
	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:421)
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:367)
	at org.terasology.engine.Terasology.main(Terasology.java:153)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
For more details, see the log files in C:\Dev\Terasology\Git\integrate\Terasology\logs\2015-08-25_00-36-15
```

I figure that's an easy fix :-)

Looks like a couple sub-menu title strings are missing in english too (under settings). And should we perhaps move the language drop-down to "Player" and rename that "Player & Language" ? It feels odd to have it under video settings :octocat: 

For future notice do we start maintaining a weblate branch here akin to how we do it with the launcher, occasionally merging translations in then updating the branch again? I see occasional merge failure emails yet.

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/250/
Hooray Jenkins reported success with all tests good!

msteiger commented:
* The merge failures are caused by amended commits. This should not be an issue as soon as this PR is merged.
* The internal server errors are caused by broken JSON. Should be fixed in Weblate 2.4.
* I'd like to work with a local branch of `develop` on the translation server instead. This ensures that commits are on top of the latest code changes. We can push/merge this branch at any time back to GitHub using the web interface.

Cervator commented:
Thanks for fix and details - merged! :-)

I figure the in-game menu is distinct from engine:menu? Noticed it shares some of the same buttons but they weren't translated - despite translations being available. Yet if they're distinct .. would we get duplicated values (but not keys) in the translation material?
